articles/types/jsdoc.md

```diff
@@ -1,3 +1,179 @@
+
+```ts
+/**
+ * Repeats a string a certain number of times.
+ * Also: the future is bright and consists of native string repetition:
+ * https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat
+ *
+ * @param {string} string  String to repeat
+ * @param {number} times   Number of times to repeat string. Integer.
+ * @see http://jsperf.com/string-repeater/2
+ */
+function repeatString(string, times) {
+  if (times === 1) {
+    return string;
+  }
+  if (times < 0) { throw new Error(); }
+  var repeated = '';
+  while (times) {
+    if (times & 1) {
+      repeated += string;
+    }
+    if (times >>= 1) {
+      string += string;
+    }
+  }
+  return repeated;
+}
+```
+
+* `@return` -- describes the return value of a callable construct (eg: a function). Type annotations and descriptions are optional.
+
+```ts
+/**
+ * Determine if the string ends with the specified substring.
+ *
+ * @param {string} haystack String to search in
+ * @param {string} needle   String to search for
+ * @return {boolean}
+ */
+function endsWith(haystack, needle) {
+  return haystack.slice(-needle.length) === needle;
+}
+```
+
+* `@class` -- tells Visual Studio to treat the next construct as a constructor. (Only available in JavaScript files)
+
+```ts
+var Cell = /** @class */ (function () {
+    function Cell() {
+    }
+    return Cell;
+}());
+```
+
+* `@augments` and `@extends` -- in JsDoc for class-like declarations, specifies which class the upcoming construct extends. Note that `@augments` and `@extends` are synonyms.
+
+```ts
+// @Filename: declarations.d.ts
+ declare class Thing<T> {
+     mine: T;
+ }
+
+// @Filename: dummy.js
+ /**
+  * @augments {Thing<string>}
+  */
+ class MyStringThing extends Thing {
+     constructor() {
+         // x is a string!
+         var x = this.mine;
+     }
+ }
+```
+
+* `@template` -- used to specify type parameters for the following construct (c.f. polymorphism in JavaScript)
+
+```ts
+/**
+ * @param {T} a
+ * @template T
+ */
+function f(a) {
+    return () => a
+}
+let n = f(1)();
+```
+
+* `@type` -- declares the type of the next construct:
+
+```ts
+/** @type {{key: "value"}} */
+var x = JSON.parse(JSON.stringify({key: "value"}));
+```
+
+* `@typedef` -- Typedef tags can be used to declare types referenced in other JsDoc comments.
+If a type is specified inline (in the same style as `@param`), that describes the shape of the `@typedef`'d name.
+Alternatively, A JsDoc beginning with a `@typedef` tag and name can be followed by `@property` and `@member` tags to describe the shape of the named type.
+
+```ts
+/**
+ * @typedef {{foo: number}} Foo1
+ */
+
+/**
+ * @typedef Foo2
+ * @property {number} foo
+ */
+
+/** @type {Foo1} */
+const x = { foo: 0 };
+```
+* `@property` and `@memeber` -- used to describe properties as part of a `@typedef` type declaration. See above for an example.
+
+## Type Syntax
+
+TODO.
```

Comment:
This will be tricky since you have to decide whether to recommend native JSDoc syntax or Typescript syntax. And it's complicated by the fact that advanced types can't be expressed in native JSDoc.

Reply:
Just linked to existing docs, adn recommended typescript